Python/marshal.c

```diff
@@ -509,17 +509,27 @@ w_complex_object(PyObject *v, char flag, WFILE *p)
             return;
         }
         W_SIZE(n, p);
-        it = PyObject_GetIter(v);
-        if (it == NULL) {
+        l = PySequence_List(v);
+        if (l == NULL) {
             p->depth--;
             p->error = WFERR_UNMARSHALLABLE;
             return;
         }
-        while ((value = PyIter_Next(it)) != NULL) {
-            w_object(value, p);
-            Py_DECREF(value);
+        for (i = 0; i < n; i++) {
+            value = PyList_GetItem(l, i);
+            value = PyMarshal_WriteObjectToString(value, Py_MARSHAL_VERSION);
```

Comment:
This will lost the identity of objects. For example, in `{(o, 1), (o, 2)}` you will get different objects after marshalling/unmarshalling.

Reply:
I'm not sure I can get the same objects back at present?
```
    def test_object_identity(self):
        o = 'test'
        obj = {(o, 1), (o, 2)}
        data = marshal.dumps(obj)
        v = marshal.loads(data)
        ids_before = {id(x) for x in obj}
        ids_after = {id(x) for x in marshal.loads(data)}
        self.assertEqual(ids_before, ids_after)

AssertionError: Items in the first set but not the second:
139864671036808
139864645838728
```
Maybe I'm misunderstanding what you mean, or that's not a good test case?

Reply:
Additional tests were added in #13736. Rebase your PR and make the tests be success.